File: Algoritmos de Ordenação/Python/quick_sort.py
def quick_sort(lista):
    divide(lista, 0, len(lista)-1)

def divide(lista, inicio, fim):
    if inicio < fim:
        
        pos_pivo = particionar(lista, inicio, fim)
        
        divide(lista, inicio, pos_pivo-1)
        divide(lista, pos_pivo+1, fim)

def particionar(lista, inicio, fim):
    pivo = lista[inicio]
    i = inicio+1
    j = fim

    executar =True
    while executar:

        while i <= j and lista[i] <= pivo:
            i = i + 1

        while lista[j] >= pivo and j >= i:
            j = j -1

        if j < i:
            executar = False
        else:
            lista[i], lista[j] = lista[j], lista[i]

    lista[inicio], lista[j] = lista[j], lista[inicio]

    return j


if __name__ == "__main__":
    from random import randint
    
    lista = [randint(-500,500) for i in range(15)]
    print(lista)

    quick_sort(lista)
    print(lista)

File: Algoritmos de Ordenação/C++/quick_sort.cpp
// PRECISA REVER

#include "extra_functions.h"

int partition(int* array, int inicio, int fim){
    int pivo = *(array+inicio);
    
    *(array+inicio) = *(array+fim);
    *(array+fim) = pivo;
    
    int i=inicio, j=inicio;
    int temp;
    
    for (j; j<fim; j++){
        
        if (*(array+j) < pivo){
            
            temp = *(array+j);
            *(array+j) = *(array+i);
            *(array+i) = temp;

            i++;
        }
    }

    *(array+fim) = *(array+i);
    *(array+i) = pivo;

    return i;
}

void divide(int* array, int inicio, int fim){
    if (inicio < fim){

        int pos_pivo = partition(array, inicio, fim);

        divide(array, inicio, pos_pivo-1);
        divide(array, pos_pivo+1, fim);

    }
}

void quick_sort(int* array, int tamanho){
    divide(array, 0, tamanho-1);
}

int main(){

    // Criação do Array de tamanho Fixo
    int tamanho = 15;
    int* array = new int[tamanho];

    // Preenchimento do array com números inteiros
    preencherArray(array, tamanho);

    printArray(array, tamanho);
    quick_sort(array, tamanho);
    printArray(array, tamanho);

    delete[] array;

    return 0;
}